### DISCUSSION TITLE
Draw Calls optimization, for the engine

### DISCUSSION THREAD

**Discussion by murataka** [cocos/cocos-engine#7630] [Ideas]
The optimization possibly needs to be done by adding or changing some features of the renderer module.

The Scene :
Labels are depth check enabled , modified shader to handle depth drawing correctly. 
The hierarchy of the node-tree is not important , just the positions are enough for calling the draw function, But still creates multiple draw calls , for same type of nodes, which can be handled in 1 draw call.
When we place many nodes , which would be optimized , possible the old 2D renderer based algorithms take place , and checks for z-orders of the node.

![Screen Shot 2020-10-29 at 11 31 05](https://user-images.githubusercontent.com/5358438/97546050-ee831680-19dc-11eb-9cd4-00871dbb4db4.png)
**this creates 2 draw calls , 1 for the labels (bitmap font) , 1 for the boxes (batched)**

![Screen Shot 2020-10-29 at 11 33 00](https://user-images.githubusercontent.com/5358438/97546045-ecb95300-19dc-11eb-8b18-4ea9266c13e6.png)

**and this creates 5 draw calls , just changed the node hierarchy , same output , same nodes, same batching optimizations .**

If there's a solution you found ( a workaround ) , let's talk about it. If not , this possibly will be a lightning performance improvement for Cocos (when z-orders of nodes do not trigger "sub draw finished", instead calling sub draw calls ,grouped to optimized batching , reagerdless of node-tree levels or z-indexes . ).

**Comment by jareguo** [comment]
@pandamicro Please look at this feedback.

**Comment by pandamicro** [comment]
I'm not sure I fully understand your question. I assume you want that the renderer can automatically improve draw call batches by reordering nodes ? If that's the case then we currently don't have any automatic order optimization algorithm.

The thing is there is always trade offs, if the renderer wants to optimize draw call, it has to know the relationship between nodes, whether they can be batched together, whether they overlaps each other, whether they need blend. All these calculations can be very complicated, for a general purpose reorder and batch algorithm from the renderer point of view. It can only be done in the CPU, with JavaScript implementation, it could be even worse than more draw calls.

I'm not saying we shouldn't do any improvement, it's just easy to see benefits and miss the trade offs. I can share what we are gonna do in 3.x

1. You can have multiple UI view that overlaps each other, each one have its own priority, so you can control the order from Canvas level.
2. You can have multiple camera to filter different node, regardless of its hierarchy relationship, camera priority is always more important
3. We will have builtin scene management in native engine, it will mainly for culling, but might have further improvements related to batching

**Reply by murataka** [reply]
Here i tried the same thing with 3D objects : 

![Screen Shot 2020-11-06 at 03 55 59](https://user-images.githubusercontent.com/5358438/98313033-fdbd2200-1fe3-11eb-85be-5115d7045e47.png)

the "mat2" materialized objects causes the draw call to x2 here(3 sorted, 7 unsorted ). 

The engine may do the very simple type of sorting, or even mapping the obects by common materials. 

e.g. : {mat2:[mobj1,mobj2],default_material:[box1,box2],multimaterial_objects:[mobi3,mobi4]}

in this scenario, mobi3 and mobi4 is mapped not to be optimized , and the renderer should calculate this grouping only when materials changed/added/deleted/ , and does not seem to cost much for js ( no need to recalculate before each render - isDirty check algorithm is enough). 


As you mentioned , the engine does not sort the nodes before rendering , which can cost too much cpu sometimes . But mapping / grouping may solve this situation.

As developers, we can try to handle this as much as possible, but this is not very efficient. For example , when designing a 3D scene , nodes must be children of others, and also we use camera culling for other purposes already.

I want to point that , the renderer does not need to know the child parent relation , but instead positions . Sorting does not affect the output. 

The only need here is, for many objects in scene draw call optimization. Think about the situation when you add 50 paired objects (1 child 1 parent each , each object has just 1 material , mate1 for parent and mate2 for child ). In this situation, the draw call is 100(not grouped) to 2 (grouped) , and this is not a rare situation (using many cloned objects in scene).

In OpenGL , which i used many years ago , i had to push - pop matrixes to move objects together , if related, and then it was a pain to calculate collisions . If you are talking about that kind of JS side calculations , you are right , this is a bit difficult. If not , mapping/ grouping by simple shaders , and then rendering in that order seems to be  really a big improvement. And i think the engine already can give us world and local coordiantes for any seperated object. So this kind of problem does not seem to exist. 

For a workaround , i will try to use more cameras for my game. I hope  i will share the result here soon.

**Reply by murataka** [reply]
![Screen Shot 2020-11-06 at 05 01 25](https://user-images.githubusercontent.com/5358438/98317203-21d13100-1fed-11eb-84b8-7f89584af861.png)

![Screen Shot 2020-11-06 at 05 02 36](https://user-images.githubusercontent.com/5358438/98317307-4c22ee80-1fed-11eb-9fec-cdecb39e6551.png)

Using 2 cameras , culling enabled for same type of materials did not work ...

**Reply by murataka** [reply]
![Screen Shot 2020-11-06 at 05 07 39](https://user-images.githubusercontent.com/5358438/98317667-00bd1000-1fee-11eb-97fd-39c1805bf84e.png)

same scene , 2 draw calls .

![Screen Shot 2020-11-06 at 05 08 22](https://user-images.githubusercontent.com/5358438/98317723-192d2a80-1fee-11eb-94b5-3388266ddc24.png)

**Reply by murataka** [reply]
Here you can test it (change the orders or child parent relation of boxes ): 
[perftest_creator2.4.3.zip](https://github.com/cocos-creator/engine/files/5498528/perftest_creator2.4.3.zip)

**Reply by pandamicro** [reply]
We are discussing different issues here actually, my previous response was for 2D objects, sorting and batching are indeed very different when we handle 2D or 3D objects.

The reason is that 2D objects need to be rendered in order so that the render order and the blend through transparent pixels is correct (objects overlaps a lot in 2D UI), so we can't simply reorder them with material hash. As for 3D objects, we do reorder them like what you suggested (in 3.0), please take a look at the [compare functions](https://github.com/cocos-creator/engine/blob/3d/cocos/core/pipeline/render-queue.ts#L42), they are invoked in the [forward stage](https://github.com/cocos-creator/engine/blob/3d/cocos/core/pipeline/forward/forward-stage.ts#L175). So our 3D objects will no longer break batches as your test in v2.4.

For 2D objects, we are also working on other optimization strategies for 3.0

**Reply by murataka** [reply]
Thank you very much for the detailed response.
I am looking forward for the new version . 

Have a nice day !
